---
title: "RunTerraWorkflow: Run workflows implemented in Terra"
author: "Sehyun Oh"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  BiocStyle::html_document:
    toc: true
    toc_float: false
    toc_depth: 3
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  comment = "#>", collapse = TRUE 
)
```

```{r echo=FALSE, eval=TRUE}
suppressPackageStartupMessages({
  library(RunTerraWorkflow)
  library(AnVIL)
})
```

# Overview
For R users with limited computing resources, we introduce RunTerraWorkflow package. 
This package allows users to run workflows implemented in [Terra](https://app.terra.bio/#), 
a cloud-based genomics platform, without writing any workflow, installing softwares, 
or managing cloud resources. Terra's computing resources are based on Google Cloud 
Platform (GCP) and to use RunTerraWorkflow, you only need to setup the Terra account 
once at the beginning.

```{r}
library(RunTerraWorkflow)
```

### Pre-requisite   
You need [Terra account setup](https://support.terra.bio/hc/en-us/articles/360034677651-Account-setup-and-exploring-Terra).
The email address linked to Terra account and the billing project name are required 
to use RunTerraWorkflow package. Here are the basic input parameters used in this example.

```{r}
accountEmail <- "shbrief@gmail.com"
billingProjectName <- "waldronlab-terra-rstudio"
```

### Input   
1-1. `cloneWorkspace` function copies the template workspace containing the workflow(s)
and the notebooks if they are available. Workspace is the main building block of 
[Terra](https://app.terra.bio/#), where different resources are delivered through.

1-2. `updateInput` function takes user's inputs. 

### Run workflow   
`launchWorkflow` function launch the workflow in Terra. You can abort the submission 
using `abortSubmission` function. 

### Result   
3-1. `monitorSubmission` function allows you to monitor the status of your workflow run.   
3-2. `listOutput` function displays the list of your workflow outputs.   
3-3. `getOutput` function allows you to download your outputs.

<br>

In this vignette, we run [Whole Metagenome Shotgun (wmgx) analysis workflow](https://github.com/biobakery/biobakery_workflows) using 
[bioBakery](https://github.com/biobakery/biobakery/wiki) tools. 
So the input value for the `analysis` argument is `bioBakery` (you
can find this information from `availableAnalysis()` function).

```{r}
availableAnalysis()
```

```{r}
analysis <- "bioBakery"
workspaceName <- "mtx_workflow_biobakery_ver3"
```


# Setup
## Clone Workspace
First, you should clone the template workspace using `cloneWorkspace` function. 
Note that you need to provide a unique name for the cloned workspace through 
`workspaceName` argument.    

If you attempt to clone the template workspace using the existing workspaceName,
you will get the below error message.

```{r}
cloneWorkspace(accountEmail, billingProjectName, workspaceName, analysis)
```

With the unique workspace name, you can successfully clone the workspace and the
function will return the name of the cloned workspace.

```{r}
cloneWorkspace(accountEmail, billingProjectName, 
               workspaceName = "microbiome", analysis)
```

```{r echo=FALSE, message=FALSE}
## Delete workspace
resp <- AnVIL::Terra()$deleteWorkspace(workspaceNamespace = billingProjectName,
                                       workspaceName = "microbiome")
rm(resp)
```


## Prepare Input
### Current input
You can review the current inputs using `currentInput` function.
```{r}
currentInput(accountEmail, billingProjectName, workspaceName)
```
<br>

`biobakery_currentInput` is a variation of `currentInput` function optimized for 
bioBakery workflow. It displays only the `inputListPath` and `inputFilePath`. File
path of your inputs stored in Google bucket (`inputFilePath`) should be provided
to the workflow in `.txt` file also saved in Google bucket (`inputListPath`).

```{r}
biobakeR::biobakery_currentInput(accountEmail, billingProjectName, workspaceName)
```


### Update input
Before launching the workflow, you should provide the correct input information 
using `updateInput` function. RunTerraWorkflow doesn't support this function yet, 
but the bioBakery-specific function is available as `biobakeR::updateInput`.

```{r}
input <- "gs://fc-07ee4ddc-5b5b-46f6-bed7-809aa14bb012/IBDMDB/ibdmdb_file_list.txt"
inputMeta <- "gs://fc-07ee4ddc-5b5b-46f6-bed7-809aa14bb012/IBDMDB/ibdmdb_demo_metadata.txt"
biobakeR::updateInput(accountEmail, billingProjectName, workspaceName,
                      InputRead1Files = input,
                      InputMetadataFile = inputMeta,
                      AdapterType = "NexteraPE",
                      ProjectName = "ibdmdb_test",
                      InputExtension = ".fastq.gz",
                      InputRead1Identifier = "_R1",
                      InputRead2Identifier = "_R2")
```



# Run bioBakery workflow
Once you cloned the template workspace and updated the input with your own data, 
you can launch the workflow using `launchWorkflow` function.

```{r}
launchWorkflow(accountEmail, billingProjectName, workspaceName)
```

## Monitor Progress
```{r}
submissions <- monitorSubmission(accountEmail, billingProjectName, workspaceName)
submissions
```

## Abort submission
To abort the most recently submitted job, you don't need to specify submissionId.
```{r}
abortSubmission(accountEmail, billingProjectName, workspaceName)
```

If the submitted job is already done, you will get the message on the result.
```{r}
abortSubmission(accountEmail, billingProjectName, workspaceName,
                submissionId = "831016f6-d9c3-4d1b-ab22-efe9fc456470")
```

If the submitted job is already aborted before done, you will get the message.
```{r}
abortSubmission(accountEmail, billingProjectName, workspaceName,
                submissionId = "c4d981e4-cafa-47cc-a0a9-0096191910c1")
```


# Result
## List Outputs 
You can check all the output files from a most recent submission with succeeded 
status (default) or a specific submission (if you specify the `submissionId` 
argument.) You can also subset outputs using `keyword` argument.

```{r}
listOutput(accountEmail, billingProjectName, workspaceName)
```

```{r}
listOutput(accountEmail, billingProjectName, workspaceName, keyword = "humann")
```


```{r echo=FALSE, eval=FALSE}
## Find a specific submissionId
jobs <- RunTerraWorkflow::monitorSubmission(accountEmail, 
                                            billingProjectName, 
                                            workspaceName)
jobs_succeed <- which(jobs$succeeded == 1)
jobs[jobs_succeed,]

submission_id <- jobs[jobs_succeed,]$submissionId[1]
submission_id
```

If your outputs include `.tsv` files, you can check the head of those files using 
`tableHead` function without downloading them. 

```{r warning=FALSE}
tableHead("HSM7J4NY_genefamilies_relab.tsv", n = 6,
          accountEmail, billingProjectName, workspaceName)   
```


## Get Outputs
`keyword` argument takes a character string containing a regular expression and as
an example here, we check all `.tsv` output of the sample, "HSM7J4NY".

```{r}
listOutput(accountEmail, billingProjectName, workspaceName, 
           keyword = "HSM7J4NY.*.tsv")
```

You can download any file using `getOutput` function. Here, we narrow down the 
download to HSM7J4NY sample's `.tsv` output files.
```{r echo=FALSE}
HSM7J4NY_dir <- "~/data2/RunTerraWorkflow/inst/extdata/outputs/HSM7J4NY/"
```

```{r eval=FALSE}
getOutput(accountEmail, billingProjectName, workspaceName, 
          keyword = "HSM7J4NY.*.tsv", dest_dir = HSM7J4NY_dir)
```

```{r}
list.files(HSM7J4NY_dir)
```

