---
title: "AnVILWorkflow usecase: Whole metagenomic shotgun sequencing data analysis using bioBakery"
author: "Sehyun Oh"
date: "`r format(Sys.time(), '%B %d, %Y')`"
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{Example 2. Metagenomic data analysis using bioBakery}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document:
    number_sections: yes
    toc: yes
    toc_depth: 4
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  comment = "#>", collapse = TRUE, fig.align = 'center',
  # eval = AnVIL::gcloud_exists(),
  eval = FALSE
)
```

# Introduction
For the introduction and setup of AnVILWorkflow package, please reference the
overview section [this vignette][].

[this vignette]: https://bioconductor.org/packages/devel/bioc/vignettes/AnVILWorkflow/inst/doc/salmon.html#overview


# Setup

## Clone workspace
```{r}
analysis <- "bioBakery"
newName <- "biobakery_test"
```

```{r}
cloneWorkspace(workspaceName = newName, 
               analysis = analysis)
```


## Prepare input

### Current input

You can review the current inputs using `currentInput` function. Below
shows all the required and optional inputs for the workflow. 
```{r}
config <- getWorkflowConfig(newName)
```

```{r}
current_input <- currentInput(newName, config)
current_input
```

<br>

### Update input
You can modify/update inputs of your workflow using `updateInput` function. To
minimize the formatting issues, we recommend to make any change in the current 
input table returned from the `currentInput` function. Under the default 
(`dry=TRUE`), the updated input table will be returned without actually
updating Terra/AnVIL. Set `dry=FALSE`, to make a change in Terra/AnVIL.

```{r}
new_input <- current_input
new_input[2,4] <- "ibdmdb_test2" # change project name
new_input

updateInput(newName, inputs = new_input) # add `dry = FALSE` to change
```


# Run workflow

You can launch the workflow using `runWorkflow()` function. 
```{r eval=FALSE}
runWorkflow(newName)
```

In case the selected workflow uses Terra's data model, you need to specify 
the `inputName` of your workflow. If you don't provide it, this function will 
return the list of input names you can use for your workflow.

```{r}
getDataTables(newName)
getDataTables(newName, table = "image")
```

```{r echo=FALSE, eval=FALSE}
## Data model version
new_input$attribute <- c("this.image_svs", "this.sampleName")
updateInput(newName, inputs = new_input, dry = FALSE) 
# runWorkflow(newName, inputName = "")
```

## Monitor progress

The last three columns (`status`, `succeeded`, and `failed`) show the 
submission and the result status. 

```{r}
submissions <- monitorWorkflow(workspaceName = newName)
submissions
```

## Abort submission

You can abort the most recently submitted job using the `stopWorkflow` 
function. You can abort any workflow that is not the most recently
submitted by providing a specific `submissionId`.

```{r}
stopWorkflow(newName)
```

# Result

The workspace `Bioconductor-Workflow-DESeq2` is the template workspace 
you cloned at the beginning using the `analysis = "salmon"` argument 
in `cloneWorkspace()` function. This template workspace has already a 
history of the previous submissions, so we will check the output examples 
in this workspace.

```{r}
submissions <- monitorWorkflow(workspaceName = newName)
submissions
```

You can check all the output files from the most recently succeeded
submission using `getOutput` function. If you specify the `submissionId` 
argument, you can get the output files of that specific submission.

```{r}
out <- getOutput(workspaceName = newName)
out
```

```{r}
getOutput(workspaceName = newName, 
          dest_dir = "biobakery_outputs",
          dry = FALSE)
```


# Session Info
```{r eval=TRUE}
sessionInfo()
```
