<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>AnVILWorkflow: Run non-R workflows implemented in Terra using R • AnVILWorkflow</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="AnVILWorkflow: Run non-R workflows implemented in Terra using R">
<meta property="og:description" content="AnVILWorkflow">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">AnVILWorkflow</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.29</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/salmon.html">Quickstart</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/shbrief/AnVILWorkflow" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>AnVILWorkflow: Run non-R workflows implemented
in Terra using R</h1>
                        <h4 data-toc-skip class="author">Sehyun Oh</h4>
            
            <h4 data-toc-skip class="date">April 13, 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/shbrief/AnVILWorkflow/blob/HEAD/vignettes/salmon.Rmd" class="external-link"><code>vignettes/salmon.Rmd</code></a></small>
      <div class="hidden name"><code>salmon.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>The <a href="https://anvilproject.org/" class="external-link">AnVIL project</a> is an
analysis, visualization, and informatics cloud-based space for data
access, sharing and computing across large genomic-related data
sets.</p>
<p>For R users with the limited computing resources, we introduce the
AnVILWorkflow package. This package allows users to run workflows
implemented in <a href="https://app.terra.bio/#" class="external-link">Terra</a> without
installing software, writing any workflow, or managing cloud resources.
Terra is a cloud-based genomics platform and its computing resources
rely on Google Cloud Platform (GCP).</p>
<p>Use of this package requires AnVIL and Google cloud computing billing
accounts. Consult <a href="https://anvilproject.org/training/guides" class="external-link">AnVIL training
guides</a> for details on establishing these accounts.</p>
<div class="section level3">
<h3 id="install-and-load-package">Install and load package<a class="anchor" aria-label="anchor" href="#install-and-load-package"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"AnVILWorkflow"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">AnVIL</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">AnVILWorkflow</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="google-cloud-sdk">Google Cloud SDK<a class="anchor" aria-label="anchor" href="#google-cloud-sdk"></a>
</h3>
<p>If you use AnVILWorkflow within Terra’s RStudio, you don’t need extra
authentication and gcloud SDK. If you use this package locally, it
requires gcloud SDK and the billing account used in Terra. You can
[install][] the gcloud sdk.</p>
<p>Check whether your system has the installation with
<code><a href="https://rdrr.io/pkg/AnVIL/man/gcloud.html" class="external-link">AnVIL::gcloud_exists()</a></code>. It should return <code>TRUE</code>
to use AnVILWorkflow package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/AnVIL/man/gcloud.html" class="external-link">gcloud_exists</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>If it returns <code>FALSE</code>, install the gcloud SDK following
this script:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"rstudio/cloudml"</span><span class="op">)</span></span>
<span><span class="fu">cloudml</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cloudml/man/gcloud_install.html" class="external-link">gcloud_install</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">## shell</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> gcloud auth login</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-terra-account">Create Terra account<a class="anchor" aria-label="anchor" href="#create-terra-account"></a>
</h3>
<p>You need <a href="https://support.terra.bio/hc/en-us/articles/360034677651-Account-setup-and-exploring-Terra" class="external-link">Terra
account setup</a>. Once you have your own Terra account, you need two
pieces of information to use AnVILWorkflow package:</p>
<ol style="list-style-type: decimal">
<li>The email address linked to your Terra account<br>
</li>
<li>Your billing project name</li>
</ol>
<p>You can setup your working environment using
<code><a href="../reference/setCloudEnv.html">setCloudEnv()</a></code> function like below. <strong>Provide the
input values with YOUR account information!</strong></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">accountEmail</span> <span class="op">&lt;-</span> <span class="st">"YOUR_EMAIL@gmail.com"</span></span>
<span><span class="va">billingProjectName</span> <span class="op">&lt;-</span> <span class="st">"YOUR_BILLING_ACCOUNT"</span></span>
<span></span>
<span><span class="fu"><a href="../reference/setCloudEnv.html">setCloudEnv</a></span><span class="op">(</span>accountEmail <span class="op">=</span> <span class="va">accountEmail</span>, </span>
<span>            billingProjectName <span class="op">=</span> <span class="va">billingProjectName</span><span class="op">)</span></span></code></pre></div>
<p>The remainder of this vignette assumes that an Terra account has been
established and successfully linked to a Google cloud computing billing
account.</p>
</div>
<div class="section level3">
<h3 id="major-steps">Major steps<a class="anchor" aria-label="anchor" href="#major-steps"></a>
</h3>
<p>Here is the table of major functions for three workflow steps -
prepare, run, and check result.</p>
<table class="table">
<colgroup>
<col width="12%">
<col width="28%">
<col width="58%">
</colgroup>
<thead><tr class="header">
<th>Steps</th>
<th>Functions</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Prepare</td>
<td><code>cloneWorkspace</code></td>
<td>Copy the template workspace</td>
</tr>
<tr class="even">
<td></td>
<td><code>updateInput</code></td>
<td>Take user’s inputs</td>
</tr>
<tr class="odd">
<td>Run</td>
<td><code>runWorkflow</code></td>
<td>Launch the workflow in Terra</td>
</tr>
<tr class="even">
<td></td>
<td><code>stopWorkflow</code></td>
<td>Abort the submission</td>
</tr>
<tr class="odd">
<td></td>
<td><code>monitorWorkflow</code></td>
<td>Monitor the status of your workflow run</td>
</tr>
<tr class="even">
<td>Result</td>
<td><code>getOutput</code></td>
<td>List or download your workflow outputs</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="example-in-this-vignette-bulk-rnaseq-analysis">Example in this vignette: bulk RNAseq analysis<a class="anchor" aria-label="anchor" href="#example-in-this-vignette-bulk-rnaseq-analysis"></a>
</h3>
<p>You can find all the available workspaces you have access to using
<code><a href="https://rdrr.io/pkg/AnVIL/man/avworkspace.html" class="external-link">AnVIL::avworkspaces()</a></code> function. Workspaces manually curated
by this package are separately checked using
<code><a href="../reference/availableAnalysis.html">availableAnalysis()</a></code> function. The values under
<code>analysis</code> column can be used for the analysis argument,
simplifying the cloning process. For this vignette, we use
<code>"salmon"</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/availableAnalysis.html">availableAnalysis</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;    analysis       workspaceNamespace                            workspaceName</span></span>
<span><span class="co">#&gt; 1 bioBakery waldronlab-terra-rstudio mtx_workflow_biobakery_version3_template</span></span>
<span><span class="co">#&gt; 2    salmon  bioconductor-rpci-anvil             Bioconductor-Workflow-DESeq2</span></span>
<span><span class="co">#&gt;           configuration_namespace              configuration_name</span></span>
<span><span class="co">#&gt; 1 mtx_workflow_biobakery_version3 mtx_workflow_biobakery_version3</span></span>
<span><span class="co">#&gt; 2         bioconductor-rpci-anvil                 AnVILBulkRNASeq</span></span>
<span><span class="co">#&gt;                                                                                              description</span></span>
<span><span class="co">#&gt; 1                                                                    Microbiome analysis using bioBakery</span></span>
<span><span class="co">#&gt; 2 Trascript quantification from RNAseq using Salmon | Differential gene expression analysis using DESeq2</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">analysis</span> <span class="op">&lt;-</span> <span class="st">"salmon"</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<div class="section level3">
<h3 id="clone-workspace">Clone workspace<a class="anchor" aria-label="anchor" href="#clone-workspace"></a>
</h3>
<div class="section level4">
<h4 id="curated-by-this-package">Curated by this package<a class="anchor" aria-label="anchor" href="#curated-by-this-package"></a>
</h4>
<p>We will refer the existing workspaces, that you have access to and
want to use for your analysis, as ‘template’ workspaces. The first step
of using this package is cloning the template workspace using
<code>cloneWorkspace</code> function. Note that you need to provide a
<strong>unique</strong> name for the cloned workspace through
<code>workspaceName</code> argument. Once you successfully clone the
workspace, the function will return the name of the cloned workspace.
For example, the successfully execution of the below script will return
<code>{YOUR_BILLING_ACCOUNT}/salmon_test</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">salmonWorkspaceName</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="st">"salmon_"</span><span class="op">)</span><span class="op">)</span> <span class="co"># unique workspace name</span></span>
<span><span class="va">salmonWorkspaceName</span></span>
<span><span class="co">#&gt; [1] "salmon_39ba8d691bb"</span></span>
<span><span class="fu"><a href="../reference/cloneWorkspace.html">cloneWorkspace</a></span><span class="op">(</span>workspaceName <span class="op">=</span> <span class="va">salmonWorkspaceName</span>, analysis <span class="op">=</span> <span class="va">analysis</span><span class="op">)</span></span>
<span><span class="co">#&gt; waldronlab-terra-rstudio/salmon_39ba8d691bb is succesfully created.</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="any-workspace-you-have-access-to">Any workspace you have access to<a class="anchor" aria-label="anchor" href="#any-workspace-you-have-access-to"></a>
</h4>
<p>If you want to clone any other workspace that you have access to but
is not curated by this pacakge, you can directly enter the name of the
target workspace as a <code>templateName</code>. For example, to clone
the <a href="https://anvil.terra.bio/#workspaces/waldronlab-terra/Tumor_Only_CNV" class="external-link">Tumor_Only_CNV</a>
workspace:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cnvWorkspaceName</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="st">"cnv_"</span><span class="op">)</span><span class="op">)</span> <span class="co"># unique workspace name</span></span>
<span><span class="va">cnvWorkspaceName</span></span>
<span><span class="fu"><a href="../reference/cloneWorkspace.html">cloneWorkspace</a></span><span class="op">(</span>workspaceName <span class="op">=</span> <span class="va">cnvWorkspaceName</span>,</span>
<span>               templateName <span class="op">=</span> <span class="st">"Tumor_Only_CNV"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="prepare-input">Prepare input<a class="anchor" aria-label="anchor" href="#prepare-input"></a>
</h3>
<div class="section level4">
<h4 id="current-input">Current input<a class="anchor" aria-label="anchor" href="#current-input"></a>
</h4>
<p>You can review the current inputs using <code>currentInput</code>
function. Below shows all the required and optional inputs for the
workflow.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">current_input</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/currentInput.html">currentInput</a></span><span class="op">(</span><span class="va">salmonWorkspaceName</span><span class="op">)</span></span>
<span><span class="va">current_input</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 4</span></span></span>
<span><span class="co">#&gt;   inputType   name                            optional attribute                </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                           <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Array[File] salmon.fastqs_1                 FALSE    <span style="color: #949494;">"</span>this.participants.fastq…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Array[File] salmon.fastqs_2                 FALSE    <span style="color: #949494;">"</span>this.participants.fastq…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> File        salmon.transcriptome_fasta      FALSE    <span style="color: #949494;">"</span>\"gs://fc-78b3af8a-a1d5…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> String      salmon.transcriptome_index_name FALSE    <span style="color: #949494;">"</span>\"new_index_name\"<span style="color: #949494;">"</span></span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="update-input">Update input<a class="anchor" aria-label="anchor" href="#update-input"></a>
</h4>
<p>You can modify/update inputs of your workflow using
<code>updateInput</code> function. To minimize the formatting issues, we
recommend to make any change in the current input table returned from
the <code>currentInput</code> function. Under the default
(<code>dry=TRUE</code>), the updated input table will be returned
without actually updating Terra/AnVIL. Set <code>dry=FALSE</code>, to
make a change in Terra/AnVIL.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">new_input</span> <span class="op">&lt;-</span> <span class="va">current_input</span></span>
<span><span class="va">new_input</span><span class="op">[</span><span class="fl">4</span>,<span class="fl">4</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"athal_index"</span></span>
<span><span class="va">new_input</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 4</span></span></span>
<span><span class="co">#&gt;   inputType   name                            optional attribute                </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                           <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Array[File] salmon.fastqs_1                 FALSE    <span style="color: #949494;">"</span>this.participants.fastq…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Array[File] salmon.fastqs_2                 FALSE    <span style="color: #949494;">"</span>this.participants.fastq…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> File        salmon.transcriptome_fasta      FALSE    <span style="color: #949494;">"</span>\"gs://fc-78b3af8a-a1d5…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> String      salmon.transcriptome_index_name FALSE    <span style="color: #949494;">"</span>athal_index<span style="color: #949494;">"</span></span></span>
<span></span>
<span><span class="fu"><a href="../reference/updateInput.html">updateInput</a></span><span class="op">(</span><span class="va">salmonWorkspaceName</span>, inputs <span class="op">=</span> <span class="va">new_input</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 9</span></span>
<span><span class="co">#&gt;  $ deleted            : logi FALSE</span></span>
<span><span class="co">#&gt;  $ inputs             :List of 4</span></span>
<span><span class="co">#&gt;   ..$ salmon.fastqs_1                : 'scalar' chr "this.participants.fastq_1"</span></span>
<span><span class="co">#&gt;   ..$ salmon.fastqs_2                : 'scalar' chr "this.participants.fastq_2"</span></span>
<span><span class="co">#&gt;   ..$ salmon.transcriptome_fasta     : 'scalar' chr "\"gs://fc-78b3af8a-a1d5-4c43-9a84-4b9036230184/athal.fa.gz\""</span></span>
<span><span class="co">#&gt;   ..$ salmon.transcriptome_index_name: 'scalar' chr "\"athal_index\""</span></span>
<span><span class="co">#&gt;  $ methodConfigVersion: int 1</span></span>
<span><span class="co">#&gt;  $ methodRepoMethod   :List of 4</span></span>
<span><span class="co">#&gt;   ..$ methodUri    : chr "dockstore://github.com%2FKayla-Morrell%2FAnVILBulkRNASeq/master"</span></span>
<span><span class="co">#&gt;   ..$ sourceRepo   : chr "dockstore"</span></span>
<span><span class="co">#&gt;   ..$ methodPath   : chr "github.com/Kayla-Morrell/AnVILBulkRNASeq"</span></span>
<span><span class="co">#&gt;   ..$ methodVersion: chr "master"</span></span>
<span><span class="co">#&gt;  $ name               : chr "AnVILBulkRNASeq"</span></span>
<span><span class="co">#&gt;  $ namespace          : chr "bioconductor-rpci-anvil"</span></span>
<span><span class="co">#&gt;  $ outputs            :List of 2</span></span>
<span><span class="co">#&gt;   ..$ salmon.salmon_index.transcriptome_index: 'scalar' chr "this.transcriptome_index"</span></span>
<span><span class="co">#&gt;   ..$ salmon.salmon_quant.quant_output       : 'scalar' chr "this.quant_output"</span></span>
<span><span class="co">#&gt;  $ prerequisites      : Named list()</span></span>
<span><span class="co">#&gt;  $ rootEntityType     : chr "participant_set"</span></span>
<span><span class="co">#&gt;  - attr(*, "class")= chr [1:2] "avworkflow_configuration" "list"</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="run-workflow">Run workflow<a class="anchor" aria-label="anchor" href="#run-workflow"></a>
</h2>
<p>You can launch the workflow using <code><a href="../reference/runWorkflow.html">runWorkflow()</a></code>
function. You need to specify the <code>inputName</code> of your
workflow. If you don’t provide it, this function will return the list of
input names you can use for your workflow.</p>
<p>Example error outputs:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/runWorkflow.html">runWorkflow</a></span><span class="op">(</span><span class="va">slamonWorkspaceName</span><span class="op">)</span></span>
<span><span class="co"># You should provide the inputName from the followings:</span></span>
<span><span class="co"># [1] "AnVILBulkRNASeq_set"</span></span>
<span><span class="co">#&gt; Error in runWorkflow(workspaceName):</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/runWorkflow.html">runWorkflow</a></span><span class="op">(</span><span class="va">salmonWorkspaceName</span>, inputName <span class="op">=</span> <span class="st">"AnVILBulkRNASeq_set"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Workflow is succesfully launched."</span></span></code></pre></div>
<div class="section level3">
<h3 id="monitor-progress">Monitor progress<a class="anchor" aria-label="anchor" href="#monitor-progress"></a>
</h3>
<p>The last three columns (<code>status</code>, <code>succeeded</code>,
and <code>failed</code>) show the submission and the result status.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">submissions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/monitorWorkflow.html">monitorWorkflow</a></span><span class="op">(</span>workspaceName <span class="op">=</span> <span class="va">salmonWorkspaceName</span><span class="op">)</span></span>
<span><span class="va">submissions</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 9</span></span></span>
<span><span class="co">#&gt;   submissionId             submitter submissionDate      status succeeded failed</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dttm&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 82565703-9db9-4b9c-8623… shbrief@… 2023-04-13 <span style="color: #949494;">22:13:29</span> Submi…         0      0</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 3 more variables: submissionRoot &lt;chr&gt;, namespace &lt;chr&gt;, name &lt;chr&gt;</span></span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="abort-submission">Abort submission<a class="anchor" aria-label="anchor" href="#abort-submission"></a>
</h3>
<p>You can abort the most recently submitted job using the
<code>stopWorkflow</code> function. You can abort any workflow that is
not the most recently submitted by providing a specific
<code>submissionId</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/stopWorkflow.html">stopWorkflow</a></span><span class="op">(</span><span class="va">salmonWorkspaceName</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Status of the submitted job (submissionId: 82565703-9db9-4b9c-8623-7fc5b2dbeb34)"</span></span>
<span><span class="co">#&gt; [1] "Workflow is succesfully aborted."</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="result">Result<a class="anchor" aria-label="anchor" href="#result"></a>
</h2>
<p>The workspace <code>Bioconductor-Workflow-DESeq2</code> is the
template workspace you cloned at the beginning using the
<code>analysis = "salmon"</code> argument in
<code><a href="../reference/cloneWorkspace.html">cloneWorkspace()</a></code> function. This template workspace has
already a history of the previous submissions, so we will check the
output examples in this workspace.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">submissions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/monitorWorkflow.html">monitorWorkflow</a></span><span class="op">(</span>workspaceName <span class="op">=</span> <span class="st">"Bioconductor-Workflow-DESeq2"</span><span class="op">)</span></span>
<span><span class="va">submissions</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 26 × 9</span></span></span>
<span><span class="co">#&gt;    submissionId            submitter submissionDate      status succeeded failed</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dttm&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> ce27a96f-f85f-4aed-9cd… kayla.mo… 2021-12-31 <span style="color: #949494;">09:33:22</span> Abort…         0      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 3b508f89-b229-42d3-867… kayla.mo… 2021-12-31 <span style="color: #949494;">09:29:47</span> Abort…         0      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> bdfc7dee-6173-4f76-867… kayla.mo… 2021-12-21 <span style="color: #949494;">13:39:38</span> Abort…         0      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 70db2183-8cad-401f-869… kayla.mo… 2021-12-21 <span style="color: #949494;">09:12:54</span> Abort…         0      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 2bfbad24-33fa-43ac-91c… kayla.mo… 2021-12-18 <span style="color: #949494;">18:15:19</span> Abort…         0      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 79cc1a62-8b90-4188-812… kayla.mo… 2021-12-16 <span style="color: #949494;">11:52:51</span> Abort…         0      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> b58888f6-085d-43d7-91b… kayla.mo… 2021-12-16 <span style="color: #949494;">11:34:06</span> Abort…         0      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> fb8e35b7-df5d-49e6-aff… kayla.mo… 2020-10-23 <span style="color: #949494;">11:12:46</span> Done           1      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> ea3f3ec1-53e5-4c66-84e… kayla.mo… 2020-10-23 <span style="color: #949494;">10:54:34</span> Done           1      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 506755cb-c318-4c2a-b81… kayla.mo… 2020-10-23 <span style="color: #949494;">10:12:02</span> Done           1      0</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 16 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 3 more variables: submissionRoot &lt;chr&gt;, namespace &lt;chr&gt;, name &lt;chr&gt;</span></span></span></code></pre></div>
<p>You can check all the output files from the most recently succeeded
submission using <code>getOutput</code> function. If you specify the
<code>submissionId</code> argument, you can get the output files of that
specific submission.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Output from the successfully-done submission</span></span>
<span><span class="va">successful_submissions</span> <span class="op">&lt;-</span> <span class="va">submissions</span><span class="op">$</span><span class="va">submissionId</span><span class="op">[</span><span class="va">submissions</span><span class="op">$</span><span class="va">succeeded</span> <span class="op">==</span> <span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getOutput.html">getOutput</a></span><span class="op">(</span>workspaceName <span class="op">=</span> <span class="st">"Bioconductor-Workflow-DESeq2"</span>,</span>
<span>                 submissionId <span class="op">=</span> <span class="va">successful_submissions</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 2</span></span></span>
<span><span class="co">#&gt;   filename           X..name.                                      </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> test_index.tar.gz  <span style="color: #949494;">"</span>,\"salmon.salmon_index.transcriptome_index\"<span style="color: #949494;">"</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> DRR016125_1.tar.gz <span style="color: #949494;">"</span>,\"salmon.salmon_quant.quant_output\"<span style="color: #949494;">"</span>       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> DRR016126_1.tar.gz <span style="color: #949494;">"</span>,\"salmon.salmon_quant.quant_output\"<span style="color: #949494;">"</span>       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> DRR016127_1.tar.gz <span style="color: #949494;">"</span>,\"salmon.salmon_quant.quant_output\"<span style="color: #949494;">"</span>       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> DRR016128_1.tar.gz <span style="color: #949494;">"</span>,\"salmon.salmon_quant.quant_output\"<span style="color: #949494;">"</span>       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> DRR016129_1.tar.gz <span style="color: #949494;">"</span>,\"salmon.salmon_quant.quant_output\"<span style="color: #949494;">"</span></span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.2 (2022-10-31)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Ventura 13.0.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] AnVILWorkflow_0.99.29 AnVIL_1.10.2          dplyr_1.1.1          </span></span>
<span><span class="co">#&gt; [4] BiocStyle_2.26.0     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.0     xfun_0.38            bslib_0.4.2         </span></span>
<span><span class="co">#&gt;  [4] purrr_1.0.1          vctrs_0.6.1          generics_0.1.3      </span></span>
<span><span class="co">#&gt;  [7] miniUI_0.1.1.1       htmltools_0.5.5      yaml_2.3.7          </span></span>
<span><span class="co">#&gt; [10] utf8_1.2.3           rlang_1.1.0          pkgdown_2.0.7       </span></span>
<span><span class="co">#&gt; [13] jquerylib_0.1.4      pillar_1.9.0         later_1.3.0         </span></span>
<span><span class="co">#&gt; [16] withr_2.5.0          DBI_1.1.3            rapiclient_0.1.3    </span></span>
<span><span class="co">#&gt; [19] glue_1.6.2           lambda.r_1.2.4       lifecycle_1.0.3     </span></span>
<span><span class="co">#&gt; [22] stringr_1.5.0        futile.logger_1.4.3  ragg_1.2.5          </span></span>
<span><span class="co">#&gt; [25] htmlwidgets_1.6.2    memoise_2.0.1        evaluate_0.20       </span></span>
<span><span class="co">#&gt; [28] knitr_1.42           fastmap_1.1.1        httpuv_1.6.9        </span></span>
<span><span class="co">#&gt; [31] curl_5.0.0           parallel_4.2.2       fansi_1.0.4         </span></span>
<span><span class="co">#&gt; [34] Rcpp_1.0.10          xtable_1.8-4         promises_1.2.0.1    </span></span>
<span><span class="co">#&gt; [37] DT_0.27              formatR_1.14         BiocManager_1.30.20 </span></span>
<span><span class="co">#&gt; [40] cachem_1.0.7         desc_1.4.2           jsonlite_1.8.4      </span></span>
<span><span class="co">#&gt; [43] mime_0.12            systemfonts_1.0.4    fs_1.6.1            </span></span>
<span><span class="co">#&gt; [46] textshaping_0.3.6    digest_0.6.31        stringi_1.7.12      </span></span>
<span><span class="co">#&gt; [49] bookdown_0.33        shiny_1.7.4          rprojroot_2.0.3     </span></span>
<span><span class="co">#&gt; [52] cli_3.6.1            tools_4.2.2          magrittr_2.0.3      </span></span>
<span><span class="co">#&gt; [55] sass_0.4.5           tibble_3.2.1         futile.options_1.0.1</span></span>
<span><span class="co">#&gt; [58] tidyr_1.3.0          pkgconfig_2.0.3      ellipsis_0.3.2      </span></span>
<span><span class="co">#&gt; [61] rmarkdown_2.21       httr_1.4.5           rstudioapi_0.14     </span></span>
<span><span class="co">#&gt; [64] R6_2.5.1             compiler_4.2.2</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Sehyun Oh.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
